File: ARE.Core/Core/GameEvent.cs
using ARE.Core.Abstractions;
using System;
using System.Collections.Generic;

namespace ARE.Core.Core
{
    public class GameEvent : IEvent
    {
        public string EventType { get; }

        public IReadOnlyDictionary<string, object> Data { get; }

        public DateTime Timestamp { get; }

        private readonly Dictionary<string, object> _data;

        public GameEvent(string eventType)
        {
            EventType = eventType;
            _data = new Dictionary<string, object>();
            Data = _data;
            Timestamp = DateTime.UtcNow;
        }

        public GameEvent Set(string key, object value)
        {
            _data[key] = value;
            return this; // fluent
        }
    }
}

File: are_engine_core/lib/src/game_event.dart
import 'interfaces.dart';

/// Default IEvent implementation.
class GameEvent implements IEvent {
  @override
  final String eventType;

  @override
  final Map<String, dynamic> data = {};

  @override
  final DateTime timestamp;

  GameEvent(this.eventType) : timestamp = DateTime.now();

  /// Set event data. Returns this for fluent chaining.
  GameEvent set(String key, dynamic value) {
    data[key] = value;
    return this;
  }
}
